### ISSUE TITLE
Engine API

### ISSUE THREAD

**Issue by CharliePoole** [nunit/docs#13]
Document the purpose and use of the Engine API in a page under technical notes. This can start from the specification page on the Dev wiki, which will have to be updated. Examples should be provided.

**Comment by CharliePoole** [comment]
This is complete and available for review along with several other API pages linked from https://github.com/nunit/docs/wiki/NUnit-Internals

**Comment by ChrisMaddock** [comment]
Looks good @CharliePoole, it'll be great to have these pages in place! Was actually a really interesting read, as someone who's not completely familiar with where responsibilities should lie within the engine.

I fixed a few minor grammar/readability things as I went - some more detailed comments below:

https://github.com/nunit/docs/wiki/NUnit-APIs

> Only one of them - the highest level - is intended for general usage. The others have very specific purposes and we recommend that they only be used for those purposes.

I'm not entirely sure what 'general usage' is here - as opposed to 'non-general usage'? What point are you trying to make here, that only the top api is treated as public-facing? Should this be 'public usage'?

https://github.com/nunit/docs/wiki/Test-Engine-API

* All links seem to be broken on this page.

> To allow future development of drivers for other frameworks and for those tests to be discovered and executed in the same way as NUnit tests.

Should this be an objective of the test engine API? Wouldn't this fall under Test Engine Driver API?

> 1. The Application base and probling path

'Probling path' - should that be 'probing path'? I'm still not too sure what that would mean in the context?

> The final and probably most frequently used method on the interface is GetRunner. It takes a TestPackage and returns an ITestRunner that is appropriate for the options specified.

Can I suggest adding a code example of this? I think this page would benefit from having a minimal example of how to programatically run tests, using the NUnit engine API. This could be combined with the example just above this block of text.

> 	Provides centralized internal trace logging for both the engine and runners (NYI)

'NYI'?

https://github.com/nunit/docs/wiki/Engine-Driver-Api

This page looks great. 👍 

https://github.com/nunit/docs/wiki/Framework-Api

Good to see an example of the reflection here - easier to follow than text for this interface perhaps!

>  Some actions take the string representation of a test filter as an argument. The NUnit Engine and Framework have shared knowledge of the format of a filter. For an empty filter (no filtering) use "".

Edit: Looks the `<filter/>` was just not rendering here. I added the backticks, that is correct usage of the framework API, isn't it?

**Comment by CharliePoole** [comment]
Thanks @ChrisMaddock - I made some updates...

https://github.com/nunit/docs/wiki/NUnit-APIs

* Revised initial text

https://github.com/nunit/docs/wiki/Test-Engine-API

* Fixed or removed links
* More precise term is "probing privatepath
* Added an example of running tests. This pointed out we need info about TestPackage, TestFilter and ITestEventHandler on this page, which I'm working on.
* The ILogging interface exists but is not yet implemented. Our current internal trace is not service-based, so the runner, engine, agent and framework all produce separate files.

https://github.com/nunit/docs/wiki/Framework-Api

Yes, `"<filter/>"` is correct use. I added back the quotes, so it's clear we are talking about a string.
